platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java

```diff
@@ -65,9 +68,82 @@ public DiagnosticToProblemListener(InternalProblemReporter problemReporter, Cont
 
     @Override
     public void report(Diagnostic<? extends JavaFileObject> diagnostic) {
+        switch (diagnostic.getKind()) {
+            case ERROR:
+                errorCount++;
+                break;
+            case WARNING:
+            case MANDATORY_WARNING:
+                warningCount++;
+                break;
+            default:
+                break;
+        }
+
         problemReporter.reporting(spec -> buildProblem(diagnostic, spec));
     }
 
+    /**
+     * This method is responsible for printing the number of errors and warnings after writing the diagnostics out to the console.
+     * This count is normally printed by the compiler itself, but when a {@link DiagnosticListener} is registered, the compiled will stop reporting the number of errors and warnings.
+     *
+     * An example output with the last two lines being the count:
+     * <pre>
+     * /.../src/main/java/Foo.java:10: error: ';' expected
+     *                     String s = "Hello, World!"
+     *                                               ^
+     * /.../Bar.java:10: warning: [cast] redundant cast to String
+     *                     String s = (String)"Hello World";
+     *                                ^
+     * 1 error
```

Comment:
Just to validate: do we have coverage when we have 0 errors and/or 1 warnings on the output?

Reply:
No, that's a good idea that we could do a combination

Reply:
Added, PTAL

Reply:
lgtm